system/jlib/jmetrics.cpp

```diff
@@ -138,35 +139,49 @@ bool MetricsManager::addMetric(const std::shared_ptr<IMetric> &pMetric)
     bool rc = false;
     std::unique_lock<std::mutex> lock(metricVectorMutex);
     std::string name = pMetric->queryName();
-    auto metaData = pMetric->queryMetaData();
-    for (auto &metaDataIt: metaData)
-    {
-        name.append(".").append(metaDataIt.value);
-    }
+    std::regex pattern("^[A-Za-z][A-Za-z0-9.]*[^.]$");
 
-    auto it = metrics.find(name);
-    if (it == metrics.end())
+    // Ensure metric name follows naming convention
+    if (std::regex_match (name,pattern))
     {
-        metrics.insert({name, pMetric});
-        rc = true;
+        auto metaData = pMetric->queryMetaData();
+        for (auto &metaDataIt: metaData)
+        {
+            name.append(".").append(metaDataIt.value);
+        }
+
+        auto it = metrics.find(name);
+        if (it == metrics.end())
+        {
+            metrics.insert({name, pMetric});
+            rc = true;
+        }
+        else
+        {
+            // If there is a match only report an error if the metric has not been destroyed in the meantime
+            auto match = it->second.lock();
+            if (match)
+            {
+#ifdef _DEBUG
+                throw MakeStringException(MSGAUD_operator, "addMetric - Attempted to add duplicate named metric with name '%s'", name.c_str());
+#else
+                OERRLOG("addMetric - Adding a duplicate named metric '%s', old metric replaced", name.c_str());
+#endif
+            }
+            else
+            {
+                rc = true;  // old metric no longer present, so it's considered unique
+            }
+            it->second = pMetric;
+        }
     }
     else
     {
-        // If there is a match only report an error if the metric has not been destroyed in the meantime
-        auto match = it->second.lock();
-        if (match)
-        {
 #ifdef _DEBUG
-            throw MakeStringException(MSGAUD_operator, "addMetric - Attempted to add duplicate named metric with name '%s'", name.c_str());
+        throw MakeStringException(MSGAUD_operator, "addMetric - Attempted to add metric with invalid name ('%s')", name.c_str());
```

Comment:
Hmmm, I don't usually throw only in debug.  I'm often trying to reproduce behavior found in non-debug builds while stepping through or increasing logging by running debug.  This could get really confusing.  Or if I didn't know how to configure something to get past something like this I would have to modify the code to keep debugging.  We do sometimes have asserts though that only happen in debug.  Might want to see what Gavin thinks.

Reply:
Only throwing in debug was a design choice made by Gavin and myself. The reason is we wanted the use to know during development that they have provided an invalid metric and to halt system execution. However, in a production environment, we did not want to halt.